File: Leetcode First 200 JS solution/5. Longest Palindromic Substring.js
/* Given a string s,
return the longest palidromic substring s
*/

var longestPalindrome = function(s) {

    // use the center spread solution
    let ans = '';
    let strLength = s.length;

    for (let i = 0; i < strLength; i++) {
        str1 = isPalindrome(i-1, i+1);
        str2 = isPalindrome(i, i+1);
        const temp = str1.length > str2.length ? str1 : str2;
        ans = temp.length > ans.length ? temp : ans;
    }
    return ans;

    function isPalindrome(leftIndex, rightIndex) {
        while (leftIndex >= 0 && rightIndex < strLength && s[leftIndex] === s[rightIndex]) {
            leftIndex -= 1;
            rightIndex += 1;
        }
        return s.slice(leftIndex + 1, rightIndex);
    }
}

// o(n^2) time complexity 
// o(n) space complexity 

var longestPalindrome = function(s) {

    // use the center spread solution
    let ans = '';
    let strLength = s.length;
    let start = 0, end = 0

    for (let i = 0; i < s.length; i++) {
        let tempStart = i;
        let tempEnd = i;

        while (s[i] === s[tempEnd + 1]) tempEnd++;

        while (tempStart >= 0 && tempEnd < s.length && s[tempStart - 1] == s[tempEnd + 1]){
            tempStart -= 1;
            tempEnd   += 1;
        }

        if (max < tempEnd - tempStart + 1) {
            start = tempStart;
            end   = tempEnd;
            max   = tempEnd - tempStart + 1;
        }
    }
    return s.substring(start, end+1);
};

// o(n^2)
// o(1) solution


// solution 3 dynamic programming

File: All Interview Package/Amazon Interviews/5. Longest Palindromic Substring.py
class Solution:
    def longestPalindrome(self, s):

        # longest palindrome substring

        ans = ''
        for i in range(len(s)):
            ans1 = self.ispalindrome(s, i, i+1)
            if len(ans1) > len(ans):
                ans = ans1
            
            ans2 = self.ispalindrome(s, i, i)
            if len(ans2) > len(ans):
                ans = ans2
            
            if len(ans) == len(s): return ans

        return ans
    
    def ispalindrome(self, s, left, right):
        while 0 <= left and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        
        return s[left+1: right]
    
    # time complexity o(n^2)
    # space o(1)



# brute force solution

def find_LPS_length(st):
    return find_LPS_length_recursive(st, 0, len(st)-1)

def find_LPS_length_recursive(st, startIndex, endIndex):
    if startIndex > endIndex:
        return 0
    
    if startIndex == endIndex:
        return 1
    
    if st[startIndex] == st[endIndex]:
        remainingLength = endIndex - startIndex - 1
        if remainingLength == find_LPS_length_recursive(st, startIndex+1, endIndex-1):
            return remainingLength + 2

    c1 = find_LPS_length_recursive(st, startIndex + 1, endIndex)
    c2 = find_LPS_length_recursive(st, startIndex, endIndex - 1)
    return max(c1, c2)

